### ISSUE TITLE
stash: The stash debug tool is unable to read, edit, or delete old stash versions

### ISSUE THREAD

**Issue by jkosh44** [MaterializeInc/materialize#21121]
### What version of Materialize are you using?

v0.65.0-dev

### What is the issue?

The stash serializes and deserializes objects using protobuf definitions found here: https://github.com/MaterializeInc/materialize/tree/main/src/stash/protos.

The stash debug tool is a tool used to help with stash debugging in customer environments. It has the following commands:
- Dump: dumps the entire stash contents into human readable JSON.
- Edit: edits an entry in the stash.
- Delete: deletes an entry in the stash.
- Upgrade check: checks if an upgrade will succeed without committing the upgrade results.

The stash debug tool has the `mz-adapter` and `mz-stash` crates as dependencies, which it uses to serialize and deserialize the stash contents. The protobuf defintions it uses is the [`objects.proto`](https://github.com/MaterializeInc/materialize/blob/main/src/stash/protos/objects.proto) file in the `mz-stash` dependency.

In order for the stash debug tool to perform the dump, edit, and delete commands, the stash debug tool needs to be the same stash version as the stash that it's inspecting.

In order for the stash debug tool to perform a meaningful upgrade check command, the stash debug tool needs to be one stash version earlier than the stash it's checking.

As a consequence, if you want to perform a meaningful upgrade check you need two versions of the stash debug tool. One version to run the upgrade check command and another version to inspect the stash contents if it fails.

**Comment by chaas** [comment]
@jkosh44 is this issue obsolete with the new persist-backed stash? What's the status of the stash debug tool after this rollout?

**Comment by jkosh44** [comment]
> @jkosh44 is this issue obsolete with the new persist-backed stash?

No, this issue also applies to the catalog-debug tool even when using the persist-backed catalog. I'll update the issue description to reflect this.

> What's the status of the stash debug tool after this rollout?

It still exists as a way to inspect and edit the storage controller's stash. Once the storage controller no longer uses the stash, then we can get rid of the stash-debug tool (and the stash itself).
